mani_skill/envs/tasks/rotate_cube.py

```diff
@@ -0,0 +1,411 @@
+
+        finger_reach_object_dist_1 = torch.norm(tip_poses[:, :3, 0] - obj_pos, p=2, dim=-1)
+        finger_reach_object_dist_2 = torch.norm(tip_poses[:, :3, 1] - obj_pos, p=2, dim=-1)
+        finger_reach_object_dist_3 = torch.norm(tip_poses[:, :3, 2] - obj_pos, p=2, dim=-1)
+        finger_reach_object_reward1 = 1 - torch.tanh(5 * finger_reach_object_dist_1)
+        finger_reach_object_reward2 = 1 - torch.tanh(5 * finger_reach_object_dist_2)
+        finger_reach_object_reward3 = 1 - torch.tanh(5 * finger_reach_object_dist_3)
+        finger_reach_object_reward = object_dist_weight * (finger_reach_object_reward1 + finger_reach_object_reward2 + finger_reach_object_reward3) / 3
+
+        # Reward for object distance
+        object_dist = torch.norm(obj_pos - goal_pos, p=2, dim=-1)
+
+        init_xyz_tensor = torch.tensor([0, 0, 0.032], dtype=torch.float, device=self.device).reshape(1, 3)
+        init_z_dist = torch.norm(init_xyz_tensor - goal_pos[...,], p=2, dim=-1)
+
+        # object_dist_reward = object_dist_weight * dt * lgsk_kernel(object_dist, scale=50., eps=2.)
+
+        object_dist_reward = 1 - torch.tanh(5 * object_dist)
+        object_init_dist_reward = 1 - torch.tanh(5 * init_z_dist)
+        object_dist_reward -= object_init_dist_reward
+
+        init_z_tensor = torch.tensor([0.032], dtype=torch.float, device=self.device).reshape(1, 1)
+        object_z_dist = torch.norm(obj_pos[..., 2:3] - goal_pos[..., 2:3], p=2, dim=-1)
+        init_z_dist = torch.norm(init_z_tensor - goal_pos[..., 2:3], p=2, dim=-1)
+        object_lift_reward = 5 * ((1 - torch.tanh(5 * object_z_dist)))
+        object_init_z_reward = 5 * ((1 - torch.tanh(5 * init_z_dist)))
+
+        object_lift_reward -= object_init_z_reward
+
+        # extract quaternion orientation
+        angles = quat_diff_rad(obj_q, goal_q)
+        object_rot_reward = -1 * torch.abs(angles)
+        pose_reward = object_dist_weight * (object_dist_reward + object_lift_reward) + object_rot_weight * object_rot_reward
+        total_reward = finger_reach_object_reward + pose_reward
+        total_reward[info["success"]] = 3
+        return total_reward
+
+    def compute_normalized_dense_reward(self, obs: Any, action: Array, info: Dict):
+        max_reward = 20
+        return self.compute_dense_reward(obs=obs, action=action, info=info) / max_reward
+
+@torch.jit.script
+def random_roll_orientation(num: int, device: torch.device) -> torch.Tensor:
+    """Returns sampled rotation around z-axis."""
+    roll = 2 * np.pi * torch.rand(num, dtype=torch.float, device=device)
+    pitch = torch.zeros(num, dtype=torch.float, device=device)
+    yaw = torch.zeros(num, dtype=torch.float, device=device)
+    euler_angles = torch.stack([roll, pitch, yaw], dim=-1)
+    rotation_matrix = euler_angles_to_matrix(euler_angles, "XYZ")
+    quat = matrix_to_quaternion(rotation_matrix)
+    return quat
+
+@torch.jit.script
+def quat_diff_rad(a: torch.Tensor, b: torch.Tensor) -> torch.Tensor:
```

Comment:
no need for you to do it, just noting to myself i will move this code to a utility file in the future as it is pretty common.